kratos/containers/pointer_vector_set.h

```diff
@@ -647,10 +648,23 @@ class PointerVectorSet final
     template <class InputIterator>
     void insert(InputIterator first, InputIterator last)
     {
-        // first sorts the input iterators and make the input unique.
-        std::sort(first, last, CompareKey());
-        auto new_last = std::unique(first, last, EqualKeyTo());
-        SortedInsert(first, new_last);
+        if constexpr(std::is_assignable_v<decltype(*std::declval<InputIterator>()), decltype(*std::declval<InputIterator>())>) {
+            // first sorts the input iterators and make the input unique if the iterators are assignable
+            std::sort(first, last, CompareKey());
+            auto new_last = std::unique(first, last, EqualKeyTo());
+            SortedInsert(first, new_last);
+        } else {
```

Comment:
@myself, this is for the case in which what i comment above is not possible ... still the comment above still holds

Reply:
> then i do not understand the role of mIds ...

`mIds` are storing the list of ids from a `PVS` when writing. So each rank will write its own local `PVS` ids which are sorted. The final list of `mIds` in the `.h5` file won't be sorted, but it will be sorted for each partition in each rank. So when you read it also, each partition will read its portion of the `mIds` from the `.h5` file whill will be an already sorted ids list. hence the `rElements.insert(rElements.end(), p_elem)` will have the hint which is valid, then `PVS` will do a push_back at the backend.

Reply:
> sorry i cannot understand this ...

This is to support iterators coming from `const std::vector<NodeType::Pointer>` for the insertion. There we need to sort and unique of the incoming iterators which is not allowed. So, you have to make a copy and do the sort and unique.  This has additional overhead, but it cannot be avoided if we allow iterators coming from immutable containers.

Reply:
yes i understand, however think at the following:

~~~cpp
std::vector my_pointer_list ....
pvs.insert(my_pointer_list.begin(), my_pointer_list.end(); //it works
//but here my_pointer_list is ordered... i would not expect this. Is that something we accept? (open question...i am not sure of what would be a standardized behaviour)
~~~

indeed

~~~cpp
const std::vector my_pointer_list .... //CONST!
//pvs.insert(my_pointer_list.begin(), my_pointer_list.end(); //will int principle not compile
pvs.insert(my_pointer_list.cbegin(), my_pointer_list.cend(); //i understand you will eventually remove the constness inside. Did i get it wrong?
//but here my_pointer_list is NOT ordered (ok, since it was const)
~~~

but now in the PVS will i have const pointers? if i do not, aren't we removing the constness wrongly?

to me this is a difficuly programming question, but i am really not clear about what will happen and OF WHAT SHOULD HAPPEN

Reply:
Hi @RiccardoRossi. This problem is fixed with the cpp standard way of doing this. The `rvalue` insertion methods are now implemented. So

```c++
std::vector my_pointer_list;
pvs.insert(my_pointer_list.begin(), my_pointer_list.end()); // this will not change the my_pointer_list, but will incur a cost of copying the data in my_pointer_list to another vector.
```

If you are sure that you don't want the `my_pointer_list` anymore, then now you can use the following:
```c++
std::vector my_pointer_list;
pvs.insert(std::move(my_pointer_list)); // this will mutate the my_pointer_list avoiding a copy operation.
```

Is this acceptable? (The benchmark shows higher performance with `rvalue` methods as well.

Reply:
yes, it also makes sense to me as it means you cannot any longer access the container from outside. 

@roigcarlo do you agree with this one?